wxDataView performance with variable line height

jensgoe opened the pull request:
This is most likely not a ready to merge pull request. It is more like a draft to start a discussion about wxDataView performance with variable line height. 

Currently the wxDataViewCtrl is unusable with the style flag wxDV_VARIABLE_LINE_HEIGHT set. A sample with approximately 100 items is enough to see laggy performance. The first commit in this PR is a quick try to reduces the calls to the method GetLineStart(). This makes performance a bit better but the result is still laggy. The next try (second commit) uses a cache for all the items line heights. This makes the DataViewCtrl really usable. 

Currently the caching is very stupid and is fully resetted when expanding or collapsing items in the tree. So i just want discuss some opportunities to get better performance into the official release. 

So suggestions are welcome.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -883,6 +888,7 @@ class wxDataViewMainWindow: public wxWindow
     bool                        m_hasFocus;
     bool                        m_useCellFocus;
     bool                        m_currentColSetByKeyboard;
+    wxHashTable                 *m_heightCache;
```
Please use `WX_DECLARE_HASH_MAP` instead of the untyped (`wxObject`-valued) legacy class.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -1724,6 +1730,7 @@ wxDataViewMainWindow::wxDataViewMainWindow( wxDataViewCtrl *parent, wxWindowID i
     m_useCellFocus = false;
     m_currentRow = (unsigned)-1;
     m_lineHeight = GetDefaultRowHeight();
+    m_heightCache = new wxHashTable();
```
This probably should be created only when `wxDV_VARIABLE_LINE_HEIGHT` is used.

vadz replied:
And it also seems to be leaked.

jensgoe replied:
Oh i mean the cache really helps currently. But you are right i am concerned too about the cache size when having a lot of items in the control. To use the cache only for the visible items might not work. It is especially necessary for all the items above the first visible item. 
Let's say i scrolled down the ScrolledWindow to y-position 100000. To repaint the control we need to know which item is the first visible item. See line 2007-2008:
```
    // compute which items needs to be redrawn
    unsigned int item_start = GetLineAt( wxMax(0,update.y) );
```
GetLineAt starts at y-position 0 and item 0, queries all cell renderers of each row for its height. If the loop reaches the target y-position (for instance 100000) it finally knows the first item to draw.. Without cache this loop takes almost 5 (five) seconds which is needed for EVERY repaint. With cache it needs a few milliseconds.

vadz replied:
Another idea to limit the cache size would be to cache only the position of each `N`-th item, with `N=100` for example, and make `GetLineAt()` smarter by only computing the offset from the last cached item before the given one. But even so it's conceivable for the cache to grow too big, so it looks like we still need some mechanism for pruning it.

vadz reviewed (dismissed):
Thanks for working on this! It would, of course, be great to optimize wxDVC in this case, but it's not really clear if it's worth having this cache if it doesn't really help (as you say).

And I'm worried about the cache growing to huge size as it's not clear what limits it. Why exactly would we want to cache the heights of items which are not currently visible or nearby? I think using a fixed size cache of a couple of screen heights (or maybe a bit more, e.g. we could also cache the first/last ones as the user can quickly go to the beginning or the end) should be sufficient and it would put to rest any concerns about the unbounded cache growth.

jensgoe commented:
Thanks for your review so far. Regarding your concern that the cache might get to big i had another idea. I commited to my branch again and tried to comment my idea.. You should look at the new class HeightCache. 
This approach is based on the fact that different item heights occurs not for every item and that item heights are reused. So having for example 1000 items in the control it is unlikely that each item has a different height. Another worst case would be if the item height changes alternating which would result in many RangeEntity elements.

jensgoe commented:
Ok i fixed the wxW style issues. What a mess...sorry for that. The mutable qualifier seems to be unnecessary, so i removed it. i developed it on the 3.0.2 branch and merged my changes to master.
I'm a bit out of practice with my C++, so like i said in my first comment it is not a ready to merge pull request. I hope others can support with testing and improvements. But this performance issue is a little annoying for me and i want it to be fixed ;-)

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -51,6 +51,8 @@
 #include "wx/stopwatch.h"
 #include "wx/weakref.h"
+#include "wx/hashmap.h"
+#include "wx/dynarray.h"
```
This really shouldn't be used in the new code, just use `wx/vector.h` and `wxVector`, which is compatible with `std::vector`, instead

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -631,6 +633,101 @@ class wxDataViewTreeNode
+struct RangeEntity
+{
+    unsigned int from;
+    unsigned int to;
```
It would be nice to have a comment saying if this is inclusive or exclusive.

P.S. After reading the rest of the code, I see it's inclusive, but I also think that it would simplify the code a bit if it were exclusive (and it would definitely be more standard for C++ where intervals are always half-open).

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -631,6 +633,101 @@ class wxDataViewTreeNode
+#include "wx/arrimpl.cpp"
+WX_DEFINE_OBJARRAY(ArrayOfRangeEntities)
+
+class RowRange
```
It's a bit confusing that this is called `RowRange` when it really seems to be a set of row ranges. I'd rename `RangeEntity` to `RowRange` (because this is what it is) and call this one `RowRanges`.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -631,6 +633,101 @@ class wxDataViewTreeNode
+public:
+    void Add(const unsigned int idx);
+    void RemoveFrom(const unsigned int idx);
+    bool Has(const unsigned int idx);
```
This method should be `const`, meaning that it should be written as

    bool Has(unsigned idx) const;

Notice that the `const` in the parameter declaration is unnecessary and is simply discarded by the C++ compiler, so it's better to remove it. However making the _method_ itself const is not unnecessary at all, as it prevents it from modifying the object and indicates to the callers that it doesn't do it.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -631,6 +633,101 @@ class wxDataViewTreeNode
+{
+public:
+    void Add(const unsigned int idx);
+    void RemoveFrom(const unsigned int idx);
```
Just curious, why is this called `RemoveFrom()` instead of `Remove()` which would be more symmetric with `Add()`?

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -631,6 +633,101 @@ class wxDataViewTreeNode
+    bool Has(const unsigned int idx);
+    unsigned int Count();
+    unsigned int CountTo(const unsigned int idx);
+    unsigned int GetSize(); // for debugging statistics
```
Just as a reminder, this method shouldn't remain in the final version of the PR.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -631,6 +633,101 @@ class wxDataViewTreeNode
+    void Add(const unsigned int idx);
+    void RemoveFrom(const unsigned int idx);
+    bool Has(const unsigned int idx);
+    unsigned int Count();
```
Maybe `CountAll()` would be better.

This also should be `const`, of course.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -631,6 +633,101 @@ class wxDataViewTreeNode
+    void RemoveFrom(const unsigned int idx);
+    bool Has(const unsigned int idx);
+    unsigned int Count();
+    unsigned int CountTo(const unsigned int idx);
```
Having a comment describing this method would be helpful, if only to indicate whether `idx` is inclusive or exclusive here.

And it should be `const` too.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -631,6 +633,101 @@ class wxDataViewTreeNode
+    void Put(const unsigned int row, const int height);
+    void Remove(const unsigned int row);
+    void Clear();
+    void LogSize(); // for debugging statistics
```
This method should be removed before integration too. I'd advise to put all this debugging code inside some `#ifdef DVC_CACHE_DEBUG` or something like this, it would make it easier to find it later.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -631,6 +633,101 @@ class wxDataViewTreeNode
+class HeightCache
+{
+public:
+    bool GetLineStart(unsigned int row, int &start);
```
This method, as well as the 2 below it, should also be const.

I also think that it would make more sense to use a simpler signature

    // Return the line starting position, in pixels, or -1 if it's not in the cache.
    int GetLineStart(unsigned row) const;

(again for this and the other 2 ones below too).

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -1724,6 +1822,7 @@ wxDataViewMainWindow::wxDataViewMainWindow( wxDataViewCtrl *parent, wxWindowID i
     m_useCellFocus = false;
     m_currentRow = (unsigned)-1;
     m_lineHeight = GetDefaultRowHeight();
+    m_rowHeightCache = new HeightCache();
```
This should only be allocated and, most importantly, used if/when necessary, i.e. when the control does have `wxDV_VARIABLE_LINE_HEIGHT` style. We definitely don't want to have all this (completely unnecessary) overhead for fixed height controls.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -1767,6 +1866,8 @@ wxDataViewMainWindow::~wxDataViewMainWindow()
 {
     DestroyTree();
     delete m_renameTimer;
+    m_rowHeightCache->Clear();
```
This isn't really needed if we delete it anyhow.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -1990,6 +2091,8 @@ void wxDataViewMainWindow::OnPaint( wxPaintEvent &WXUNUSED(event) )
         return;
     }
 
+//    m_rowHeightCache->LogSize();
```
To be removed.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -3192,6 +3356,303 @@ int wxDataViewMainWindow::GetLineHeight( unsigned int row ) const
 
+void RowRange::Add(const unsigned int idx)
+{
+    RangeEntity *rng = NULL;
```
Raw pointers should be avoided in C++ code, instead of declaring it here you should just use a reference below:

    RangeEntity& rng = m_ranges[rngIdx];

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -3192,6 +3356,303 @@ int wxDataViewMainWindow::GetLineHeight( unsigned int row ) const
+
+    size_t count = m_ranges.GetCount();
+    size_t rngIdx = 0;
+    while (rngIdx < count)
```
This loop should be written as

    for ( unsigned rngIdx = 0; rngIdx < count; ++rngIdx )

as this is exactly what it does and seeing `while` here is confusing as you expect something more complicated than this, precisely because the loop is not written using just a simple `for`.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -3192,6 +3356,303 @@ int wxDataViewMainWindow::GetLineHeight( unsigned int row ) const
+
+void RowRange::RemoveFrom(const unsigned int idx)
+{
+    RangeEntity *rng = NULL;
```
Same remarks as above: please prefer references to pointers and `for` to `while` in this function too.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -3192,6 +3356,303 @@ int wxDataViewMainWindow::GetLineHeight( unsigned int row ) const
+        if (prevRng != NULL && prevRng->to >= rng->to)
+        {
+            m_ranges.RemoveAt(rngIdx);
+            count = m_ranges.GetCount();
```
Could just do `count--`.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -3192,6 +3356,303 @@ int wxDataViewMainWindow::GetLineHeight( unsigned int row ) const
+{
+    RangeEntity *rng = NULL;
+
+    size_t count = m_ranges.GetCount();
```
This should be const, i.e.

    const size_t count = m_ranges.GetCount();

to avoid changing it accidentally later and to indicate that it won't be change to the reader.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -3192,6 +3356,303 @@ int wxDataViewMainWindow::GetLineHeight( unsigned int row ) const
+
+    size_t count = m_ranges.GetCount();
+    size_t rngIdx = 0;
+    if (idx > 0)
```
This check, `idx - 1` below and `idx + 1` in the loop all look like a code smell to me. I'm not sure at all if there are no off-by-1 errors here and it would be really better to define precisely what the index is because it's not clear which row range exactly does it correspond to.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -3192,6 +3356,303 @@ int wxDataViewMainWindow::GetLineHeight( unsigned int row ) const
+    size_t count = m_ranges.GetCount();
+    for (size_t rngIdx = 0; rngIdx < count; rngIdx++)
+    {
+        rng = m_ranges[rngIdx];
```
Here pointers are not used, but we copy the value (granted, a small one) unnecessarily, so a reference would still be better:

    const RangeEntity& rng = m_ranges[rngIdx];

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -3192,6 +3356,303 @@ int wxDataViewMainWindow::GetLineHeight( unsigned int row ) const
+{
+    RangeEntity rng;
+    size_t count = m_ranges.GetCount();
+    for (size_t rngIdx = 0; rngIdx < count; rngIdx++)
```
Couldn't (and then shouldn't) this method use binary search?

vadz reviewed (dismissed):
First of all, really sorry for taking so long to get back to this, but (hopefully) better late than never.

Anyhow, the changes globally look fine to me, although there are plenty of small comments about the details. There is one blocker too: all this code shouldn't be used at all unless `wxDV_VARIABLE_LINE_HEIGHT` is used.

I think it would be also better to put it into `include/wx/private/heightcache.h` or something like this and write some unit tests using the classes in this header as it's just not obvious that there are no off-by-1 or other errors here and having tests would be very reassuring.

Last, but probably most importantly, I'm still not really sure if it's actually better than the simple "direct" cache. This one still can consume O(N) (where `N` is the number of items) memory as you already pointed out with the example of odd/even rows having 2 distinct heights. Is it really sufficiently better than a fixed size direct cache? It's definitely much less simple...

Finally, I'm going to apply (with some minor changes) your first commit, reducing the number of `GetLineStart()` calls, as this definitely doesn't have any ill effects and does help, even if only a little. But I'm still not sure about the main/rest of the patch.

jensgoe commented:
thanks for the valuable feedback. i am going to apply your suggestions very soon and update this pull requests.

jensgoe commented:
@vadz i continued working on this. i am not finished with your review comments yet. But i moved the implementation to a separate module in the private section and added some tests. The CI is complaining about some linker errors. Do i have to modify the makefiles for eaach platform myself? Or are they generated? I am not sure to which library i link against. First i added it to the adv lib because dataview is located there. Now it is in the base lib. What do you prefer?

vadz commented:
Makefiles are generated, but unfortunately the documentation for this process in `docs/contributing/how-to-add-files-to-build-system.md` is probably out of date. Briefly, you need to add the new files to `build/files` (as you did), then run `build/upmake` and then run `bakefile_gen` from bakefile 0.2.11 release. You don't seem to have done the latter as `Makefile.in` was not updated and hence the new file is not built in Unix builds (except for those using CMake).

Concerning which library it should be in, I'd say that putting it in "adv" is fine, as it's a private class, we can always move it to "base" or, more likely, "core", in the future if it's needed elsewhere, but for now I don't think we should link all console applications with this code without any reason for doing it.

TIA!

jensgoe commented:
Do u have a clue why i get those linker errors in the build for the unit tests? I already moved my test to the gui_test package because this is linked against wx_adv in that my implementation of smartrowheightcache is located.

vadz commented:
You need to use `WXDLLIMPEXP_ADV` in the declarations of the classes that you want to use from outside the library they're defined in, otherwise they're internal to the shared library in which they're defined.

I didn't really look at the code itself yet, but I did notice one thing: please don't use `OBJARRAY` macro, just use `wxVector<>` instead. TIA!

jensgoe commented:
@vadz Hi. Now this branch has probably a good state to review it again.
i changed the RowRange behaviour to be half opened like you suggested. The dataview only uses the height cache if the VARIABLE_LINE_HEIGHT flag is used. 
Currently i kept the method signature for the Get...-methods that returns boolean results. To return an invalid row with a value of -1 for instance made some trouble with "unsigned int" types. By the way this seems to be an issue right now. See commit b256ef4. i had some infinite loops due to overflows. Maybe this sould be fixed as a separate pull request. 
Anyway if my changes fits all the requested contribution rules some day i can open a new branch and pull request to have a single commit for all my changes or you do a squashed merge.

jensgoe commented:
Is there any chance to get this pull request moving forward? The last problem was only some linker errors in ms build that i didn't have locally. 
In the meanwhile some conflicts has shown up.

vadz commented on include/wx/private/smartrowheightcache.h:
```diff
@@ -0,0 +1,142 @@
+// Purpose:     Smart height cache of rows in a dataview
+// Author:      Jens Goepfert
+// Created:     2018-03-06
+// Copyright:   (c) 2018 Vadim Zeitlin <vadim@wxwidgets.org>
```
Not really. Please assign copyright to "wxWidgets team" if you don't care about it.

jensgoe replied:
i updated this

vadz commented on include/wx/private/smartrowheightcache.h:
```diff
@@ -0,0 +1,142 @@
+///////////////////////////////////////////////////////////////////////////////
+// Name:        wx/private/smartrowheightcache.h
```
I'm a bit uncomfortable about using "smart" in this file name. First because it looks like it's something that Julian Smart would have written and second because it's not really that smart in the first place. Let's just call it `rowheightcache`.

vadz commented on include/wx/private/smartrowheightcache.h:
```diff
@@ -0,0 +1,142 @@
+    Adds a row index to this group by adding it to an existing RowRange
+    or by creating a new one.
+    */
+    void Add(const unsigned int row);
```
Please don't use top level `const`, it's useless and inconsistent with the rest of wxWidgets code.

jensgoe replied:
i updated this

vadz commented on include/wx/private/smartrowheightcache.h:
```diff
@@ -0,0 +1,142 @@
+Use bisect algorithm in combination with GetLineStart() to
+find the appropriate item
+*/
+class WXDLLIMPEXP_ADV HeightCache
```
This class doesn't need to be exported from the shared library as it's only used inside it.

jensgoe replied:
after updating to current master i changed this to WXDLLIMPEXP_CORE. Somehow i need this to get the tests running. Otherwise i got linker errors.

vadz commented on include/wx/private/smartrowheightcache.h:
```diff
@@ -0,0 +1,142 @@
+
+
+#endif // _WX_PRIVATE_SMARTROWHEIGHTCACHE_H_
+
```
Very minor, but please remove extra blank line at the end of file.

jensgoe replied:
done

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -820,7 +821,7 @@ class wxDataViewMainWindow: public wxWindow
     int GetLineStart( unsigned int row ) const;  // row * m_lineHeight in fixed mode
     int GetLineHeight( unsigned int row ) const; // m_lineHeight in fixed mode
-    int GetLineAt( unsigned int y ) const;       // y / m_lineHeight in fixed mode
+    int GetLineAt( int y ) const;                // y / m_lineHeight in fixed mode
```
I know you mentioned this in a PR comment, but I still don't really understand why did this need to be changed. Is `GetLineAt()` supposed to be called with non-positive values of `y`? IMO this shouldn't be the case, i.e. we need to find out where does this happen and fix the caller. And if it doesn't happen, then this change should be just reverted because it doesn't change anything, of course.

FWIW I wouldn't be against using `int` -- but then we need to do it everywhere, not just in a single function here. And to avoid changing too many things at once, for now I'd rather avoid changing this.

jensgoe replied:
allright i changed this back.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -1980,6 +1990,11 @@ wxDataViewMainWindow::~wxDataViewMainWindow()
 {
     DestroyTree();
     delete m_renameTimer;
+    if (GetOwner()->HasFlag(wxDV_VARIABLE_LINE_HEIGHT))
```
I'd check `m_rowHeightCache` directly. We know that it needs to be deleted if it had been allocated, even if `wxDV_VAR_LINE_HEIGHT` flag is somehow not set any longer.

Note that the same comment applies to all the other tests of this flag in the code below, testing directly for `m_rowHeightCache` is both simpler, shorter and more robust.

jensgoe replied:
done

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -3275,6 +3302,8 @@ void wxDataViewMainWindow::SendSelectionChangedEvent( const wxDataViewItem& item
 
 void wxDataViewMainWindow::RefreshRow( unsigned int row )
 {
+    if (row == ((unsigned)-1))
```
Do we really want to allow calling `RefreshRow(-1)`? Where does it happen?

I.e. the question is whether we shouldn't assert that the row is valid (`< total`) instead of silently returning here?

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -3377,47 +3422,68 @@ int wxDataViewMainWindow::GetLineStart( unsigned int row ) const
-int wxDataViewMainWindow::GetLineAt( unsigned int y ) const
+int wxDataViewMainWindow::GetLineAt( int y ) const
 {
+    if (y < 0) return -1;
```
Same comment as for `RefreshRow()` above: IMO we should assert here, not return silently.

Also, unrelated, but please don't write `if` statements on a single line. If nothing else, this makes it impossible to put a breakpoint on `return` statement in most (all?) debuggers and this would be a very useful thing to do in this particular case in order to find out when this happens.

jensgoe replied:
removed it. it is probably not necessary anymore

vadz commented on tests/smartrowheightcache/smartrowheightcachetest.cpp:
```diff
@@ -0,0 +1,326 @@
+
+#include "wx/private/smartrowheightcache.h"
+
+// ----------------------------------------------------------------------------
```
No need for this block...

vadz commented on tests/smartrowheightcache/smartrowheightcachetest.cpp:
```diff
@@ -0,0 +1,326 @@
+// test class
+// ----------------------------------------------------------------------------
+
+class SmartRowHeightCacheTestCase : public CppUnit::TestCase
```
I won't ask you to change this if you don't want to, but it would be great if new tests could be written without the compatibility CppUnit macros and just use CATCH (which is used now under the hood even for such tests anyhow) directly.

This would make things simpler as you can just write `TEST_CASE() { ... your test code here ... }` without bothering with defining any classes, suites etc and more convenient to use during run-time as you would be able to select just a particular test to run.

For an example of a test using CATCH directly, see `tests/html/htmprint.cpp` for example.

jensgoe replied:
it is a CATCH test now

vadz reviewed (commented):
I'm really sorry for missing your previous comment saying that the new version was ready for review!

I did look at it now and it looks good to me, but I'd like to tweak a few more things (file name, test for the cache directly, ...) and, perhaps most importantly, understand whether we want to allow calling `RefreshRow()` and `GetLineAt()` with invalid parameters. IMHO we shouldn't be doing this but I could be wrong, I don't know this code well. What do you think?

jensgoe committed:
Using a cache for the items height if using wxDV_VARIABLE_LINE_HEIGHT

review comments
- copyright notice
- modules renamed
- always delete m_rowHeightCache if it was set

jensgoe committed:
refactored tests to run with CATCH

jensgoe committed:
updated make files

jensgoe committed:
args for GetLineAt keeps unsigned int

vadz commented:
Thanks for the update, I'll try to merge this (really) soon, i.e. before 3.1.2.

vadz commented:
Sorry for only getting back to it now, but I've finally tried actually testing this and unfortunately it just doesn't work at all for me, if I apply the following patch to the sample:
{{{
#!diff
diff --git a/samples/dataview/dataview.cpp b/samples/dataview/dataview.cpp
index 51273a6106..b1a23d862f 100644
--- a/samples/dataview/dataview.cpp
+++ b/samples/dataview/dataview.cpp
@@ -773,7 +773,7 @@ void MyFrame::BuildDataViewCtrl(wxPanel* parent, unsigned int nPanel, unsigned l
             wxASSERT(!m_ctrl[2]);
             wxDataViewListCtrl* lc =
                 new wxDataViewListCtrl( parent, wxID_ANY, wxDefaultPosition,
-                                        wxDefaultSize, style );
+                                        wxDefaultSize, style | wxDV_VARIABLE_LINE_HEIGHT);
             m_ctrl[2] = lc;

             MyListStoreDerivedModel* page2_model = new MyListStoreDerivedModel();
}}}

I get an immediate crash due to a repeating assertion failure, as the item used in `GetLineAt()` is invalid.

I'm leaving my changes at #1053 and I'll try to return to this, but if you could please look at it and fix the bug, it would be great.

vadz commented:
I think we can close this in favour of #1053 now as it has more fixes.

Anil8753 commented:
@jensgoe @vadz 
I tried building wx3.1.4 on Mac with the following macros:
wxHAS_GENERIC_DATAVIEWCTRL
wxUSE_GENERICDATAVIEWCTRL

I know I should not use generic DVC on mac instead should use the native one, but I have to use generic DVC on mac.

I am getting the following errors:

**Undefined symbols for architecture x86_64:
  "HeightCache::GetLineInfo(unsigned int, int&, int&)", referenced from:
      wxDataViewMainWindow::GetLineAt(unsigned int) const in datavgen.o
  "HeightCache::GetLineStart(unsigned int, int&)", referenced from:
      wxDataViewMainWindow::GetLineStart(unsigned int) const in datavgen.o
  "HeightCache::GetLineHeight(unsigned int, int&)", referenced from:
      wxDataViewMainWindow::GetLineAt(unsigned int) const in datavgen.o
      wxDataViewMainWindow::GetLineStart(unsigned int) const in datavgen.o
      wxDataViewMainWindow::GetLineHeight(unsigned int) const in datavgen.o
  "HeightCache::Put(unsigned int, int)", referenced from:
      wxDataViewMainWindow::QueryAndCacheLineHeight(unsigned int, wxDataViewItem) const in datavgen.o
  "HeightCache::Clear()", referenced from:
      wxDataViewMainWindow::ClearRowHeightCache() in datavgen.o
  "HeightCache::Remove(unsigned int)", referenced from:
      wxDataViewMainWindow::ItemDeleted(wxDataViewItem const&, wxDataViewItem const&) in datavgen.o
      wxDataViewMainWindow::DoItemChanged(wxDataViewItem const&, int) in datavgen.o
      wxDataViewMainWindow::Expand(unsigned int) in datavgen.o
      wxDataViewMainWindow::Collapse(unsigned int) in datavgen.o
  "HeightCache::GetLineAt(int, unsigned int&)", referenced from:
      wxDataViewMainWindow::GetLineAt(unsigned int) const in datavgen.o
  "HeightCache::~HeightCache()", referenced from:
      wxDataViewMainWindow::~wxDataViewMainWindow() in datavgen.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)**

jensgoe commented:
i try to figure out how. But i have no mac environment to test.
